bench_pm/main.c

```diff
@@ -0,0 +1,226 @@
+
+#ifdef MODULE_PERIPH_LLWU
+/* Platform specific configuration for testing Kinetis low leakage wake up module (LLWU) */
+#ifndef LLWU_WAKE_PIN
+/* This macro should correspond to the pin used in the GPIO_WAKE_PIN macro */
+#if defined(BOARD_FRDM_KW41Z)
+#define LLWU_WAKE_PIN       LLWU_WAKEUP_PIN_PTC4
+#elif defined(BOARD_FRDM_K22F)
+#define LLWU_WAKE_PIN       LLWU_WAKEUP_PIN_PTC1
+#elif defined(BOARD_FRDM_K64F)
+#define LLWU_WAKE_PIN       LLWU_WAKEUP_PIN_PTC6
+#endif
+#endif /* LLWU_WAKE_PIN */
+#endif /* MODULE_PERIPH_LLWU */
+
+#ifndef TEST_PIN_ON
+#ifdef LED0_ON
+#define TEST_PIN_ON         LED0_ON
+#define TEST_PIN_OFF        LED0_OFF
+#else
+#define TEST_PIN_ON
+#define TEST_PIN_OFF
+#endif /* LED0_ON */
+#endif /* TEST_PIN_ON */
+
+#ifndef TEST_ISR_PIN_ON
+#ifdef LED1_ON
+#define TEST_ISR_PIN_ON     LED1_ON
+#define TEST_ISR_PIN_OFF    LED1_OFF
+#else
+#define TEST_ISR_PIN_ON
+#define TEST_ISR_PIN_OFF
+#endif /* LED1_ON */
+#endif /* TEST_ISR_PIN_ON */
+
+#ifdef MODULE_PERIPH_RTT
+#define TICKS_TO_WAIT       (10 * RTT_FREQUENCY)
+#endif /* MODULE_PERIPH_RTT */
+
+static volatile unsigned busy = 0;
+
+#ifdef MODULE_PERIPH_RTT
+static void cb_rtt(void *arg)
+{
+    (void)arg;
+    TEST_ISR_PIN_ON;
+    DEBUG("RTT IRQ\n");
+
+    busy = 0;
+}
+#endif /* MODULE_PERIPH_RTT */
+
+#ifdef MODULE_PERIPH_LLWU
+static void cb_llwu(void *arg)
+{
+    (void)arg;
+    TEST_ISR_PIN_ON;
+    DEBUG("llwu pin\n");
+
+    busy = 0;
+}
+#endif /* MODULE_PERIPH_LLWU */
+
+#ifdef MODULE_PERIPH_GPIO_IRQ
+static void cb_gpio(void *arg)
+{
+    (void)arg;
+    TEST_ISR_PIN_ON;
+    DEBUG("gpio pin\n");
+
+    busy = 0;
+}
+#endif /* MODULE_PERIPH_GPIO_IRQ */
+
+int main(void)
+{
+    TEST_PIN_ON;
+    puts("\nRIOT power consumption and wake timing test application");
+
+#ifdef MODULE_PERIPH_LLWU
+    /* Kinetis specific */
+    puts("Enable LLWU wake up from RTC");
+    llwu_wakeup_module_enable(LLWU_WAKEUP_MODULE_RTC_ALARM);
+    puts("Enable LLWU IRQ on PTC4 (SW4) falling");
+    gpio_init(GPIO_WAKE_PIN, GPIO_IN);
+    llwu_wakeup_pin_set(LLWU_WAKE_PIN, LLWU_WAKEUP_EDGE_FALLING, cb_llwu, NULL);
+#endif /* MODULE_PERIPH_LLWU */
+
+#ifdef MODULE_PERIPH_GPIO_IRQ
+    puts("Enable GPIO IRQ on PTC4 (SW4) falling");
+    int res = gpio_init_int(GPIO_WAKE_PIN, GPIO_IN_PU, GPIO_FALLING, cb_gpio, NULL);
+    if (res != 0) {
+        printf("!! gpio_init_int: %d\n", res);
+    }
+#endif /* MODULE_PERIPH_GPIO_IRQ */
+
+#ifdef MODULE_PERIPH_RTT
+    puts("Initializing the RTT driver");
+    rtt_init();
+#endif /* MODULE_PERIPH_RTT */
+
+#if ENABLE_CLKOUT
+    /* Kinetis specific, clock monitor via CLKOUT on pin PTB3 */
+    puts("Enable CLKOUT on PTB3");
+    PORTB->PCR[3] = PORT_PCR_MUX(4);
+    /* Select which clock to output */
+    SIM->SOPT2 = (SIM->SOPT2 & ~SIM_SOPT2_CLKOUTSEL_MASK) | SIM_SOPT2_CLKOUTSEL(2);
+    /* Use a logic analyzer or oscilloscope to look at the signal */
+#endif
+
+#if COMA_MODE
+    puts("Coma mode, going to pm_set(0), not coming back");
```

Comment:
I've had some boards that made it really difficult to re-attach a debugger when going to deep sleep too fast.
Could we maybe guard this, either by a second delay or using something like ```while (gpio_read(BUTTON1)) {}```?

Reply:
> some boards that made it really difficult to re-attach a debugger

Like, the only way was to do "while true; do make -C examples/hello-world flash; done" and hammering the reset button, hoping to hop into the short time from reset to CPU stop...